### PR TITLE
Encapsulate all CAMP calls in a single place. Remove code repetition.

### PR DESCRIPTION
**Pull request by andreaturli**
@neykov as it is based on your #436 could you please make sure I'm not missing anything here?

### CODE REVIEW

File: core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java [hunk 1]
```diff
@@ -82,41 +100,90 @@
...
+
     @SuppressWarnings("unchecked")
-    public static <T extends Application> T createUnstarted(ManagementContext mgmt, AssemblyTemplate at) {
-        CampPlatform camp = getCampPlatform(mgmt).get();
-        AssemblyTemplateInstantiator instantiator;
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, Reader yaml, BrooklynClassLoadingContext loader) {
+        return (EntitySpec<? extends Application>) createSpec(mgmt, yaml, loader);
+    }
+
+    private static <T, SpecT> SpecT createEntitySpec(String symbolicName, CampPlatform camp,
```
Comment (neykov) [inline core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java]: No need for the generics, can return `EntitySpec<? extends Application>`

File: usage/camp/src/main/java/io/brooklyn/camp/brooklyn/YamlLauncherAbstract.java [hunk 2]
```diff
@@ -79,8 +80,8 @@ public Application launchAppYaml(String url) {
 
     public Application launchAppYaml(String url, boolean waitForTasksToComplete) {
         try {
-            Reader input = Streams.reader(new ResourceUtils(this).getResourceFromUrl(url));
-            Application app = launchAppYaml(input, waitForTasksToComplete);
```
Comment (neykov) [inline usage/camp/src/main/java/io/brooklyn/camp/brooklyn/YamlLauncherAbstract.java]: This is the wrong method to update, replace the one that was called here. For a global check if the patch was successful check for references like `registerDeploymentPlan`, `instantiate`, they should we restricted to the `EntityManagementUtils` class only (outside of camp projects).

**Comment by neykov** [comment]
I am getting the following error when trying to start an app from yaml:
`com.sun.proxy.$Proxy78 cannot be cast to brooklyn.entity.Application`
with yaml
```
location: broo
services:
- type: brooklyn.entity.webapp.tomcat.TomcatServer
```

Something wrong with my build?

**Comment by andreaturli** [comment]
thanks @neykov for your comments. I've addressed them but I think there are still problems 

Failed tests:
  ApplicationResourceTest.testDeployApplicationYaml:236 expected [simple-app-yaml] but found [Application (p6lRzMdf)]
  ApplicationResourceTest.testReferenceCatalogEntity:249 response is Client response status: 500 expected [true] but found [false]

not sure what is causing them to fail. can you @neykov help on that?

**Comment by andreaturli** [comment]
@neykov looks like the problem is with 'AssemblyTemplateSpecInstantiator.createNestedSpec' used by 'EntityManagementUtils.createEntitySpec' 
I know you spent some time on it, so maybe it is easier for you to understand the issue

File: usage/camp/src/main/java/io/brooklyn/camp/brooklyn/YamlLauncherAbstract.java [hunk 3]
```diff
@@ -130,4 +126,13 @@ public void destroyAll() {
         }
     }
 
+   private Application createFromUrl(Reader reader) {
```
Comment (neykov) [inline usage/camp/src/main/java/io/brooklyn/camp/brooklyn/YamlLauncherAbstract.java]: The argument is no longer a URL, rename method. Also the TODO below is not needed in this case.

File: core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java [hunk 4]
```diff
@@ -82,41 +100,90 @@
...
+    public static EntitySpec<? extends Application> createEntitySpec(ManagementContext mgmt, Reader yaml, BrooklynClassLoadingContext loader) {
+        return (EntitySpec<? extends Application>) createSpec(mgmt, yaml, loader);
+    }
+
+    private static EntitySpec<? extends Application> createEntitySpec(String symbolicName, CampPlatform camp,
+                                                                      DeploymentPlan plan, BrooklynClassLoadingContext loader) {
+        AssemblyTemplate at = registerDeploymentPlan(camp, plan, loader);
+        AssemblyTemplateInstantiator instantiator = newInstantiator(at);
+        if (instantiator instanceof AssemblyTemplateSpecInstantiator) {
+            return (EntitySpec<? extends Application>) ((AssemblyTemplateSpecInstantiator)instantiator).createNestedSpec(at, camp, loader, MutableSet.of(symbolicName));
```
Comment (neykov) [inline core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java]: Make sure that `createNestedSpec` is called only from `BasicBrooklynCatalog` code (as the `createEntitySpec` wrapping it). Atm it's used for creating apps, thus the error.
Also rename the methods coming from `BasicBrooklynCatalog` to indicate that they are used for catalog items.

File: core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java [hunk 5]
```diff
@@ -82,41 +100,90 @@
...
+    }
+
+    public static AbstractBrooklynObjectSpec<?, ?> createSpec(ManagementContext mgmt, String yaml, BrooklynClassLoadingContext loader) {
+        return createSpec(mgmt, new StringReader(yaml), loader);
+    }
+
+    public static AbstractBrooklynObjectSpec<?, ?> createSpec(ManagementContext mgmt, Reader yaml, BrooklynClassLoadingContext loader) {
+        Preconditions.checkNotNull(mgmt, "mgmt");
+        Preconditions.checkNotNull(mgmt, "yaml");
+        Preconditions.checkNotNull(mgmt, "loader");
```
Comment (neykov) [inline core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java]: Copy-paste discrepancy

File: core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java [hunk 6]
```diff
@@ -82,41 +100,90 @@
...
+        DeploymentPlan plan = parseDeploymentPlan(camp, yaml);
+        return createSpec(null, mgmt, plan, loader);
+    }
+
+    private static AbstractBrooklynObjectSpec<?, ?> createSpec(String symbolicName, ManagementContext mgmt,
+                                                               DeploymentPlan plan, BrooklynClassLoadingContext loader) {
+        if (isPolicyPlan(plan)) {
+            return createPolicySpec(plan, loader);
+        } else {
+            return createEntitySpec(symbolicName, getCampPlatform(mgmt), plan, loader);
```
Comment (neykov) [inline core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java]: Here is where things go sideways. The method `createEntitySpec` called here is coming from `BasicBrooklynCatalog`, calling into `createNestedSpec`. You should instead use https://github.com/neykov/incubator-brooklyn/commit/55305ea9bf05598a8ac2c811691008818edbf321#diff-b713fc5de648db4fede03fa71cfd1c6fR315 which calls into `AssemblyTemplateSpecInstantiator.createSpec`.

Comment (andreaturli) [inline core/src/main/java/brooklyn/management/internal/EntityManagementUtils.java]: Notice that currently BasicBrooklynCatalog looks like https://github.com/apache/incubator-brooklyn/blob/1b0037b609b4ad11a803522e696d6bfcac00f57f/core/src/main/java/brooklyn/catalog/internal/BasicBrooklynCatalog.java#L299-L325
so I think here it is doing the same

**Comment by ahgittin** [comment]
great idea here @andreaturli .  i'm not sure `EntityManagementUtils` is the natural home but then i'm not sure what is.

it will probably need redoing however given changes since this PR and also in #585.  tedious, but i don't think there are any difficult changes.

**Comment by andreaturli** [comment]
Closing as #760 is what we want to merge

**Pull request closed without merging**